control/discovery.py

```diff
@@ -0,0 +1,958 @@
+               self.reserved2 + self.subnqn + self.reserved3 + \
+               self.nvmeof_attributes + self.power_state_attributes + \
+               self.vendor_specific
+
+
+class CqeSetFeature:
+    def __init__(self):
+        # DWORD0
+        self.dword0 = bytearray(4)
+        # DWORD1
+        self.dword1 = bytearray(4)
+        # SQ head pointer
+        self.sq_head_ptr = bytearray(2)
+        # SQ identifier
+        self.sq_id = bytearray(2)
+        # command identifier
+        self.cmd_id = bytearray(2)
+        # status field
+        self.status = bytearray(2)
+
+    def compose_reply(self):
+        return self.dword0 + self.dword1 + self.sq_head_ptr + \
+               self.sq_id + self.cmd_id + self.status
+
+class NVMeGetLogPage:
+    def __init__(self):
+        # generation counter
+        self.genctr = bytearray(8)
+        # number of records
+        self.numrec = bytearray(8)
+
+        #record format
+        self.recfmt = bytearray(2)
+        self.reserved = bytearray(1006)
+
+    def compose_short_reply(self):
+        return self.genctr + self.numrec
+
+    def compose_data_reply(self):
+        return self.genctr + self.numrec + self.recfmt + self.reserved
+
+class DiscoveryLogEntry:
+    def __init__(self):
+        # transport type
+        self.trtype = bytearray(1)
+        # adress family
+        self.adrfam = bytearray(1)
+        # subsystem type
+        self.subtype = bytearray(1)
+        # transport requirement
+        self.treq = bytearray(1)
+        # port ID
+        self.port_id = bytearray(2)
+        # controller ID
+        self.controller_id = bytearray(2)
+        # admin max SQ size
+        self.asqsz = bytearray(2)
+        self.reserved1 = bytearray(22)
+        # transport service indentifier
+        self.trsvcid = bytearray(32)
+        self.reserved2 = bytearray(192)
+        # NVM subsystem qualified name
+        self.subnqn = bytearray(256)
+        # Transport address(TRADDR)
+        self.traddr = bytearray(256)
+        # Transport specific address subtype
+        self.tsas = bytearray(256)
+
+    def compose_reply(self):
+        return self.trtype + self.adrfam + self.subtype + \
+               self.treq + self.port_id + self.controller_id + \
+               self.asqsz + self.reserved1 + self.trsvcid + \
+               self.reserved2 + self.subnqn + self.traddr + self.tsas
+
+class DiscoveryService:
+    """Implements discovery controller.
+
+    Response discover request from initiator.
+
+    Instance attributes:
+        version: Discovery controller version
+        config: Basic gateway parameters
+        logger: Logger instance to track discovery controller events
+        omap_name: OMAP object name
+        ioctx: I/O context which allows OMAP access
+        discovery_addr: Discovery controller addr which allows initiator send command
+        discovery_port: Discovery controller's listening port
+    """
+
+    BDEV_PREFIX = "bdev_"
```

Comment:
Python already provides a [`socketserver.TCPServer`](https://docs.python.org/3/library/socketserver.html#asynchronous-mixins) (with threaded or forked implementations) which takes care of the basic streaming server duties.

Reply:
Hey @CongMinYin, sorry for my late response! Unless you really want to practice some Python networking, personally I wouldn't really mess with that. As I mentioned before, network servers rank top-1 as vector attack, and offloading this concern to a 3rd party code will allow us to focus more on what's specific about NVMe disc.

I don't get the "persistent connection" requirement. Is this a functional requirement (does NVMe disc require that?), or is it performance-wise?

Due to the Global Interpreter Lock, Python is not fully multi-threading, and there are multiple solutions to connection multiplexing (forking, threading, `selectors`, coroutines, ...), but I don't think we should dig that deeper given that there are well-adopted Python frameworks for that:
* The already mentioned Python stdlib [`sockerserver`](https://docs.python.org/3/library/socketserver.html),
* The newer stdlib [`asyncio`](https://docs.python.org/3/library/asyncio-stream.html), which relies on coroutines,
* 3rd party libraries, like [`twisted`](https://docs.twisted.org/en/twisted-18.9.0/core/howto/servers.html) or [`tornado`](https://www.tornadoweb.org/en/stable/).